internal/net/http/transport/roundtrip.go

```diff
@@ -41,50 +42,66 @@ func NewExpBackoff(opts ...Option) http.RoundTripper {
 	return e
 }
 
-func (e *ert) RoundTrip(req *http.Request) (res *http.Response, err error) {
+// RoundTrip round trip the HTTP request and return the response.
+// If backoff is not set, the default roundTrip implementation will be used.
+// The default roundTrip implementation round trip the request, and return any round trip error occurred.
+// It returns errors.ErrTransportRetryable to indicate if the request is consider as retryable.
+func (e *ert) RoundTrip(req *http.Request) (*http.Response, error) {
 	if e.bo == nil {
 		return e.roundTrip(req)
 	}
-	var r interface{}
-	r, err = e.bo.Do(req.Context(), func() (interface{}, error) {
+
+	r, err := e.bo.Do(req.Context(), func() (interface{}, error) {
 		return e.roundTrip(req)
```

Comment:
![badge](https://img.shields.io/badge/review-ask-blueviolet.svg?style=flat-square)
there's no need to check the error is `errors.ErrTransportRetryable` or not?
i think if a returned error is not `errors.ErrTransportRetryable`, it should escape the `Do` block immediately.

Reply:
but i'm not sure about a correct behavior of RoundTripper. please consider about the problem above.

Reply:
I am not sure too. Seems the retry mechanism is not implemented yet? 
I am thinking that we cannot stop the retry mechanism in backoff unless we cancel the context.
@kpango what do you think?

Reply:
ErrTransportRetryable is new feature of this PR I think we should strictly check error is Retryable or not and if err is not ErrTransportRetryable we should pass error outside of backoff

Reply:
@kpango do you mean we need to update backoff implementation to check the retryable status?

Reply:
no I mean you should check error is retryable or not in the Do function

Reply:
ErrTransportRetryable is only for transport

Reply:
how can we control the backoff retry logic here?
https://github.com/vdaas/vald/blob/master/internal/backoff/backoff.go#L80

Reply:
no you shouldn't change backoff.go you can change this line
```suggestion
	res, err := e.roundTrip(req)
	if err != nil {
		if err == errors.ErrTransportRetryable {
			return nil,  errors.ErrTransportRetryable
		}
		// write err detailt to outside err variable eg. boerr
		return nil, nil // e.bo.Do returns nil, nil but we have to check boerr
	}
	return res, nil
```

Reply:
Thinking we can add a new function to backoff to handle the retry mechanism.

Reply:
@kpango @rinx  I have pushed the changes here. Thank you :)
I will implement test case if this changes is good.
https://github.com/vdaas/vald/pull/589/commits/7a186a806d4400b2092b7add9f7fbdd5710c1354

Reply:
We can add a new error called `ErrBackoffRetryable` and check the error in backoff implementation. 
If you support my suggestion I will create a new proposal on SET project board first :)

Reply:
thank you for your suggestion, please make design proposal for backoff package.
I agree with your opinion, I think it's better to handle by boolean for example
change Do interface to 
```
Do(ctx context.Context, fn func() (val interface{}, retryable bool, err error))
```

Reply:
thank you for your comment. I think this is better maybe 🤔
`Do(ctx context.Context, fn func() (val interface{}, status backoff.Status, err error))`

Reply:
I think it's boring to define specific type for backoff enable/disable because there is only 2 choice
if we define like my idea users can use it easly.
```
r, err := e.bo.Do(req.Context(), func() (interface{}, bool, error) {
	res, err :=  e.roundTrip(req)
        return res, err == errors. ErrTransportRetryable, err
```
or 
```
r, err := e.bo.Do(req.Context(), func() (interface{}, bool, error) {
	res, err :=  e.roundTrip(req)
        return res, err != nil, err
```

Reply:
by the way, please send design proposals we should discussed there

Reply:
you are right, I will create a proposal first and lets discuss the design there.
In this PR I will not include that new feature to the backoff first, and I will implement new test code base on the current implement.

Reply:
@kpango @rinx FYI
https://github.com/vdaas/vald/projects/3#card-42651099